File: src/infrastructure/memory/memorySupport.ts
import type { Clock, IdGenerator, UnitOfWork } from '../../application/ports/repositories'
import type { Id } from '../../domain/entities/types'
import type { Snapshotable } from './memoryRepositories'

/** أدوات مساعدة للاختبار: وحدة العمل والمعرّفات والساعة. */

/**
 * وحدة عمل ذرّية فعليًا: تلتقط لقطة من كل مستودع قبل البدء،
 * وتستعيدها كاملة عند أي فشل — فلا تبقى نصف دفعة.
 *
 * تثبت حالة spec/06: «انقطاع أثناء حفظ دفعة ⇒ صفر أو كامل الدفعة».
 */
export class MemoryUnitOfWork implements UnitOfWork {
  /**
   * كل مستودع يلتقط نوع لقطة مختلفًا، والوحدة تخزّنها وتعيدها كما هي
   * بلا أن تعرف شكلها — لذلك `unknown` هو النوع الصحيح هنا.
   */
  constructor(private readonly stores: readonly Snapshotable<unknown>[]) {}

  async run<T>(work: () => Promise<T>): Promise<T> {
    const snapshots = this.stores.map((s) => s.snapshot())
    try {
      return await work()
    } catch (error) {
      // الاستعادة بالترتيب العكسي ليس مهمًا هنا لأن اللقطات مستقلة،
      // لكن الاستعادة **كلها** مهمة: مستودع واحد لم يُستعد = نصف دفعة
      this.stores.forEach((store, i) => store.restore(snapshots[i]))
      throw error
    }
  }
}

/** وحدة عمل بلا تراجع — للاختبارات التي لا تحتاج محاكاة الفشل. */
export class PassthroughUnitOfWork implements UnitOfWork {
  async run<T>(work: () => Promise<T>): Promise<T> {
    return work()
  }
}

export class SequentialIdGenerator implements IdGenerator {
  private counters = new Map<string, number>()
  next(prefix: string): Id {
    const n = (this.counters.get(prefix) ?? 0) + 1
    this.counters.set(prefix, n)
    return `${prefix}-${String(n).padStart(6, '0')}`
  }
}

export class FixedClock implements Clock {
  constructor(private readonly iso: string) {}
  nowIso(): string {
    return this.iso
  }
}

File: native-app/app/src/commonMain/kotlin/app/masroufy/memory/MemorySupport.kt
package app.masroufy.memory

import app.masroufy.core.Id
import app.masroufy.port.Clock
import app.masroufy.port.IdGenerator
import app.masroufy.port.UnitOfWork

/** أدوات مساعدة للاختبار — نقل `memorySupport.ts`. */

/**
 * مستودع يقدر يرجّع حالته لِما كانت عليه — هو اللي بيخلي `MemoryUnitOfWork`
 * ذرّية فعلًا مش اسمًا (spec/06: «انقطاع وسط حفظ دفعة ⇒ صفر أو كامل الدفعة»).
 * الكيانات في كوتلن ثابتة (data class)، فنسخة الخريطة نفسها كافية.
 */
interface Snapshotable {
    fun snapshot(): Any

    fun restore(state: Any)
}

/** وحدة عمل ذرّية فعليًا: لقطة من كل مستودع قبل البدء، واستعادة كاملة عند أي فشل. */
class MemoryUnitOfWork(private val stores: List<Snapshotable>) : UnitOfWork {
    override suspend fun <T> run(work: suspend () -> T): T {
        val snapshots = stores.map { it.snapshot() }
        try {
            return work()
        } catch (error: Throwable) {
            // الاستعادة **كلها** مهمة: مستودع واحد ما اترجعش = نص دفعة
            stores.forEachIndexed { i, store -> store.restore(snapshots[i]) }
            throw error
        }
    }
}

/** وحدة عمل من غير تراجع — للاختبارات اللي مش محتاجة محاكاة الفشل. */
class PassthroughUnitOfWork : UnitOfWork {
    override suspend fun <T> run(work: suspend () -> T): T = work()
}

class SequentialIdGenerator : IdGenerator {
    private val counters = mutableMapOf<String, Int>()

    override fun next(prefix: String): Id {
        val n = (counters[prefix] ?: 0) + 1
        counters[prefix] = n
        return "$prefix-${n.toString().padStart(6, '0')}"
    }
}

class FixedClock(private val iso: String) : Clock {
    override fun nowIso(): String = iso
}
